src/mavsdk/core/mavlink_mission_transfer_server.cpp

```diff
@@ -301,4 +333,363 @@ void MavlinkMissionTransferServer::ReceiveIncomingMission::callback_and_reset(Re
     _done = true;
 }
 
+MavlinkMissionTransferServer::SendOutgoingMission::SendOutgoingMission(
+    Sender& sender,
+    MavlinkMessageHandler& message_handler,
+    TimeoutHandler& timeout_handler,
+    uint8_t type,
+    const std::vector<ItemInt>& items,
+    double timeout_s,
+    ResultCallback callback,
+    uint8_t target_system_id,
+    uint8_t target_component_id,
+    bool debugging) :
+    WorkItem(sender, message_handler, timeout_handler, type, timeout_s, debugging),
+    _items(items),
+    _callback(callback),
+    _target_system_id(target_system_id),
+    _target_component_id(target_component_id)
+{
+    _message_handler.register_one(
+        MAVLINK_MSG_ID_MISSION_REQUEST,
+        [this](const mavlink_message_t& message) { process_mission_request(message); },
+        this);
+
+    _message_handler.register_one(
+        MAVLINK_MSG_ID_MISSION_REQUEST_INT,
+        [this](const mavlink_message_t& message) { process_mission_request_int(message); },
+        this);
+
+    _message_handler.register_one(
+        MAVLINK_MSG_ID_MISSION_ACK,
+        [this](const mavlink_message_t& message) { process_mission_ack(message); },
+        this);
+}
+
+MavlinkMissionTransferServer::SendOutgoingMission::~SendOutgoingMission()
+{
+    _message_handler.unregister_all(this);
+    _timeout_handler.remove(_cookie);
+}
+
+void MavlinkMissionTransferServer::SendOutgoingMission::start()
+{
+    std::lock_guard<std::mutex> lock(_mutex);
+
+    _started = true;
+
+    for (unsigned i = 0; i < _items.size(); ++i) {
+        if (_items[i].seq != i) {
+            LogWarn() << "Invalid sequence";
+            callback_and_reset(Result::InvalidSequence);
+            return;
+        }
+    }
+
+    if (_items.size() > 0) {
```

Comment:
Note that this is different from the Mission Transfer Client, specifically we do allow send operations when the mission count is 0, in other words the ground station can request a mission download, and get an answer which is an empty mission.

Reply:
Actually I think you want to keep this specific check. This is doing an integrity check on non-empty missions to ensure that only one mission item is marked as "current". The concern is the `if(fum_currents != 1)` check below on line 394 will fail if the mission count is 0, so the point is that we only do these checks if the mission is non-empty. Does that make sense?

Reply:
Actually I'm sorry I just realized the spirit of your question is more to the point: why are we validating and erroring out on a mission that's being requested for download? Presumably the mission was already validated on upload to the vehicle. I agree this should be removed.

- [ ] Remove